crates/isograph_compiler/src/watch.rs

```diff
@@ -44,18 +52,50 @@ pub async fn handle_watch_command<TOutputFormat: OutputFormat>(
                         );
                         watcher.stop();
                         (rx, watcher) = create_debounced_file_watcher(&state.config);
-                        WithDuration::new(|| state.compile())
+                        WithDuration::new(|| {
+                            let source_files = SourceFiles::read_all(&mut state.db, &state.config)?;
+                            let result =
+                                compile::<TOutputFormat>(&state.db, &source_files, &state.config);
+                            state.source_files = Some(source_files);
+                            result
+                        })
                     } else if changes.len() < MAX_CHANGED_FILES {
                         info!("{}", "File changes detected. Starting to compile.".cyan());
-                        WithDuration::new(|| state.update(&changes))
+                        WithDuration::new(|| {
+                            if let Some(source_files) = state.source_files.as_mut() {
+                                source_files.read_updates(
+                                    &mut state.db,
+                                    &state.config,
+                                    &changes,
+                                )?;
+                                compile::<TOutputFormat>(&state.db, source_files, &state.config)
+                            } else {
+                                let source_files =
+                                    SourceFiles::read_all(&mut state.db, &state.config)?;
+                                let result = compile::<TOutputFormat>(
+                                    &state.db,
+                                    &source_files,
+                                    &state.config,
+                                );
+                                state.source_files = Some(source_files);
+                                result
+                            }
+                        })
                     } else {
                         info!(
                             "{}",
                             "Too many changes. Starting a full compilation.".cyan()
```

Comment:
Perhaps this should go... re-reading everything means that we call `db.set` on everything and throw all intermediate results away, but the whole point of pico is that it can handle whatever we throw at it and gracefully recompute only what's needed. Beforehand, I would've believed that we had some poorly tested logic, but with pico, I think that this is unnecessary caution.

Anyway, could totally be separate PR